Add SIMD acceleration for Matrix4x4.Invert #34394

eanova opened the pull request:
Fix for #34394.

Added SIMD hardware acceleration support to the Matrix4x4.Invert function.

Commit:
Add SIMD acceleration for Matrix4x4.Invert #34394

Fix for #34394.

Added SIMD hardware acceleration support to the Matrix4x4.Invert function.

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Numerics/Matrix4x4.cs:
```diff
@@ -1311,8 +1311,180 @@ public static Matrix4x4 CreateReflection(Plane value)
-        public static bool Invert(Matrix4x4 matrix, out Matrix4x4 result)
+        public static unsafe bool Invert(Matrix4x4 matrix, out Matrix4x4 result)
         {
+            if (Sse.IsSupported)
```
Was this ported from the DirectX Math implementation?

If so, we need to document that and add some third party notices.

eanova replied:
I can add a comment and update the third party notices page.   One other thing I wanted to add is the test case for when the determinant is zero, currently that path is not being tested.

Commit:
Add link to source and update THIRD-PARTY-NOTICES.TXT

Added the link to Microsoft/DirectXMath source code and appended license to THIRD-PARTY-NOTICES.TXT

Commit:
Add test for non-invertable matrix.

Given a Matrix4x4 of only rank 3 test to see the matrix is non-invertable.

Commit:
Typo fixed in new test case

Commit:
Fixed formating for test matrix.

watfordgnf commented on src/libraries/System.Private.CoreLib/src/System/Numerics/Matrix4x4.cs:
```diff
@@ -1311,8 +1311,183 @@ public static Matrix4x4 CreateReflection(Plane value)
+                result.M42 = row4.GetElement<float>(1);
+                result.M43 = row4.GetElement<float>(2);
+                result.M44 = row4.GetElement<float>(3);
+            }
```
Is this missing a `return true` or an `else` wrapping the fallback?

eanova replied:
Good catch!  The other functions use the same style, but I suppose the compiler would have caught this error if the fallback was wrapped.

Commit:
Fix for missing return statement.

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Numerics/Matrix4x4.cs:
```diff
@@ -1311,8 +1311,183 @@ public static Matrix4x4 CreateReflection(Plane value)
-        public static bool Invert(Matrix4x4 matrix, out Matrix4x4 result)
+        public static unsafe bool Invert(Matrix4x4 matrix, out Matrix4x4 result)
         {
+            if (Sse.IsSupported)
```
Given how big the individual implementations are, it might be good to move them to different methods:
```csharp
[MethodImpl(MethodImplOptions.AggressiveInlining)]
public static unsafe bool Invert(Matrix4x4 matrix, out Matrix4x4 result)
{
    if (Sse.IsSupported)
    {
        return SseImpl();
    }

    return SoftwareFallback();

    static bool SseImpl(Matrix4x4 matrix, out Matrix4x4 result)
    {
    }

    static bool SoftwareFallback(Matrix4x4 matrix, out Matrix4x4 reuslt)
    {
    }
}
```

We've done this for various other methods where the different paths are this big.

eanova replied:
I've broken up the two implementations.

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Numerics/Matrix4x4.cs:
```diff
@@ -1311,8 +1311,183 @@ public static Matrix4x4 CreateReflection(Plane value)
+                Vector128<float> row4 = Sse.LoadVector128(&matrix.M41);
+
+                // Transpose the matrix
+                Vector128<float> vTemp1 = Sse.Shuffle(row1, row2, 0x44);
```
nit: For the various shuffle constants, it would be beneficial to annotate them like:
```csharp
Sse.Shuffle(row1, row2, 0x44); // _MM_SHUFFLE(1, 0, 1, 0);
```

This allows, at a glance, for the arrangement to be determined. Alternatively, it could be done like:
```csharp
Sse.Shuffle(row1, row2, 0b_01_00_01_00);
```
Which is a bit easier to separate out/update the arrangements for (but not as clear as the comment).

eanova replied:
The first one is definitely more readable I think.  I added the comments to all Shuffles and Permutes.

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Numerics/Matrix4x4.cs:
```diff
@@ -1311,8 +1311,183 @@ public static Matrix4x4 CreateReflection(Plane value)
+                row3 = Sse.Shuffle(vTemp3, vTemp4, 0x88);
+                row4 = Sse.Shuffle(vTemp3, vTemp4, 0xDD);
+
+                Vector128<float> V00 = Sse.Shuffle(row3, row3, 0x50);
```
The DXMath implementation defines a `Permute` helper function that uses `Avx.Permute` or `Sse.Shuffle` for when `op1`/`op2` are the same.

We could trivially define the following and get the same benefit:
```csharp
[MethodImpl(MethodImplOptions.AggressiveInlining)]
static Vector128<float> Permute(Vector128<float> value, byte control)
{
    if (Avx.IsSupported)
    {
        return Avx.Permute(value, control);
    }

    Debug.Assert(Sse.IsSupported);
    return Sse.Shuffle(value, value, control);
}
```

eanova replied:
Added back the Permute function using both Avx.Permute and Sse.Shuffle

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Numerics/Matrix4x4.cs:
```diff
@@ -1311,8 +1311,183 @@ public static Matrix4x4 CreateReflection(Plane value)
+                V02 = Sse.Shuffle(row3, row1, 0xDD);
+                V12 = Sse.Shuffle(row4, row2, 0x88);
+
+                D0 = Sse.Subtract(D0, Sse.Multiply(V00, V10));
```
Might be worth a comment that we deviate here and only use `Subtract(op, Multiply(op, op))`. It probably needs further discussion/consideration if we can have the path deviate on hardware with FMA support as it will result in different answers in some cases.

eanova replied:
Added explanation to why this expansion was used.

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Numerics/Matrix4x4.cs:
```diff
@@ -1311,8 +1311,183 @@ public static Matrix4x4 CreateReflection(Plane value)
+
+                // Get the determinant
+                vTemp2 = row1;
+                Vector128<float> vTemp = Sse.Multiply(C0, vTemp2);
```
You should be able to use `Vector4.Dot(C0.AsVector4(), vTemp2.AsVector4()).AsVector128()` so we can use `DPPS` on hardware with it available. `AsVector4` and `AsVector128` should be zero cost conversions.

eanova replied:
The result of Vector4.Dot is a float,  I used that to pack a Vector128<Float>.Create for use later in the method.  Also updated the Determinant check to use only the float from the Dot method.

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Numerics/Matrix4x4.cs:
```diff
@@ -1311,8 +1311,183 @@ public static Matrix4x4 CreateReflection(Plane value)
+                    return false;
+                }
+
+                Vector128<float> ones = Vector128.Create(1.0f, 1.0f, 1.0f, 1.0f);
```
nit: It's better to use `Vector128.Create(1.0f)`

eanova replied:
Updated that line and the new test case I added.

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Numerics/Matrix4x4.cs:
```diff
@@ -1311,8 +1311,183 @@ public static Matrix4x4 CreateReflection(Plane value)
+                row3 = Sse.Multiply(C4, vTemp);
+                row4 = Sse.Multiply(C6, vTemp);
+
+                result.M11 = row1.GetElement<float>(0);
```
nit: It would be better to use 4x `Sse.StoreVector128`

eanova replied:
Sse.Store(&result.M41, row4);

Whenever I try doing this I get the following error.

Error CS0212 You can only take the address of an unfixed expression inside of a fixed statement initializer

Seems to be a consequence of using the out keyword.  I can however create a temporary Matrix4x4 and store into that, and then copy over the new reference, but it's more expensive.

tannergooding replied:
The following should work and would avoid the 16 individual stores:
```csharp
Unsafe.SkipInit(out result);
ref Vector128<float> vResult = ref Unsafe.As<Matrix4x4, Vector128<float>>(ref result);

vResult = row1;
Unsafe.Add(ref vResult, 1) = row2;
Unsafe.Add(ref vResult, 2) = row3;
Unsafe.Add(ref vResult, 3) = row4;
```

tannergooding replied:
There is an issue requesting we add `ref` overloads for `Load/Store` here: https://github.com/dotnet/runtime/issues/36182

I'll cross reference the PR as another place it would have made things simpler.

eanova replied:
I was able to use your suggestion in a local project I'm using for testing and benchmarks using 5.0-preview3.  Unfortunately I'm having issues now with building locally where the coreclr and mono Private.CoreLib does not recognized Unsafe,  even though it's clearly in the shared project items.

tannergooding commented:
Overall, LGTM. I just had a few code cleanup and codegen quality nits.
It would be good to post the benchmark numbers here for reference as well.

Commit:
Add suggested fixes to Matrix4x4.Invert

Update containing all suggested fixes.

eanova committed:
Merge branch 'master' into SIMDMatrix4x4

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Numerics/Matrix4x4.cs:
```diff
@@ -1305,13 +1306,204 @@ public static Matrix4x4 CreateReflection(Plane value)
+            return true;
+        }
+
+        public static bool SoftwareFallback(Matrix4x4 matrix, out Matrix4x4 result)
```
This can't be public and both should likely be `static local functions` as they will only ever be used from within the publicly exposed API.

Commit:
Added missing using statement

Added missing using statement for Internal.Runtime.CompilerServices.Unknown static object.

turbedi commented on src/libraries/System.Private.CoreLib/src/System/Numerics/Matrix4x4.cs:
```diff
@@ -1305,13 +1307,204 @@ public static Matrix4x4 CreateReflection(Plane value)
+            }
+
+            // Create Vector128<float> copy of the determinant and invert them.
+            Vector128<float> ones = Vector128.Create(1.0f, 1.0f, 1.0f, 1.0f);
```
```suggestion
            Vector128<float> ones = Vector128.Create(1.0f);
```

eanova replied:
Thanks!

turbedi commented on src/libraries/System.Private.CoreLib/src/System/Numerics/Matrix4x4.cs:
```diff
@@ -1305,13 +1307,204 @@ public static Matrix4x4 CreateReflection(Plane value)
+
+            // Create Vector128<float> copy of the determinant and invert them.
+            Vector128<float> ones = Vector128.Create(1.0f, 1.0f, 1.0f, 1.0f);
+            Vector128<float> vTemp = Vector128.Create(det, det, det, det);
```
```suggestion
            Vector128<float> vTemp = Vector128.Create(det);
```

eanova commented:
``` ini

BenchmarkDotNet=v0.12.1, OS=Windows 10.0.18363.836 (1909/November2018Update/19H2)
Intel Core i9-9900K CPU 3.60GHz (Coffee Lake), 1 CPU, 16 logical and 8 physical cores
.NET Core SDK=5.0.100-preview.3.20216.6
  [Host]     : .NET Core 5.0.0 (CoreCLR 5.0.20.21406, CoreFX 5.0.20.21406), X64 RyuJIT
  DefaultJob : .NET Core 5.0.0 (CoreCLR 5.0.20.21406, CoreFX 5.0.20.21406), X64 RyuJIT


```
|                       Method |     Mean |    Error |   StdDev |
|----------------------------- |---------:|---------:|---------:|
|        OriginalDotNetInverse | 41.18 ns | 0.288 ns | 0.269 ns |
| FinalSIMDInverseWithSSEStore | 31.30 ns | 0.023 ns | 0.020 ns |

These are the results corresponding to the current commit,  which uses the CompilerServices Unknown class to store the final output using SSE Store intrinsics.  I followed the the benchmarking guidelines in the docs to generate the tests.

Commit:
Use abbreviated constructor for Vector128

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Numerics/Matrix4x4.cs:
```diff
@@ -1305,13 +1307,204 @@ public static Matrix4x4 CreateReflection(Plane value)
+
+
+
+        private static unsafe bool SseImpl(Matrix4x4 matrix, out Matrix4x4 result)
```
This and the other should still be made `static local functions` as I called out here: https://github.com/dotnet/runtime/pull/36323#discussion_r426401821

Otherwise they may conflict with other methods we also accelerate.

tannergooding replied:
Everything thing else LGTM and I believe it is otherwise ready to merge

eanova replied:
Sorry, wasn't familiar with local functions.  I thought you were pointing out the public modifier instead.  Fixed!

tannergooding replied:
No worries, I should have clarified 😄

Commit:
Moved implementations into local functions

Moved the SSE implementation and SoftwareFallback to local functions of Invert.

tannergooding commented:
CC. @pgovind if you want to give this a pair of eyes before it gets merged

eanova commented:
Thanks for all your help Tanner.  Really appreciate it!
